### ISSUE TITLE
4 XMX8 Matrix test timeout with "[InstCombine] Canonicalise ADD+GEP"

### ISSUE THREAD

**Issue by jsji** [intel/llvm#11538]
**Describe the bug**
Instcombine unable to optimize out load of matrix causing timeout of matrix tests.

Timed Out Tests (4):
SYCL :: Matrix/Legacy/XMX8/element_wise_all_ops_int8.cpp
SYCL :: Matrix/XMX8/element_wise_all_ops.cpp
SYCL :: Matrix/XMX8/element_wise_all_ops_int8.cpp
SYCL :: Matrix/XMX8/element_wise_all_ops_int8_packed.cpp

**To Reproduce**
Cherry-pick e13bed4c5f3544c076ce57e36d9a11eefa5a7815 and build 

**Environment (please complete the following information):**
CI Matrix test on Intel Arc A-series

**Additional context**

```
  --- f2.ll 2023-10-13 07:44:54.812588744 -0700
  +++ b2.ll 2023-10-13 07:44:39.180398753 -0700
  @@ -2,6 +2,13 @@
   ; Function Attrs: convergent norecurse nounwind
   define weak_odr dso_local spir_kernel void @_ZTSZZ19matrix_verify_logicIaLm16ELm16EEvN4sycl3_V15queueER10big_matrixIT_XT0_EXT1_EERNS1_8nd_rangeILi2EEEiENKUlRNS1_7handlerEE_clESB_E12logic_matrix(ptr addrspace(1) noundef align 1 %_arg_accA, ptr noundef byval(%"class.sycl::_V1::range") align 8 %_arg_accA1, ptr noundef byval(%"class.sycl::_V
   entry:
  +  %_arg_accA2.ascast = addrspacecast ptr %_arg_accA2 to ptr addrspace(4)
  +  %_arg_accA3.ascast = addrspacecast ptr %_arg_accA3 to ptr addrspace(4)
  +  %agg.tmp5.sroa.0.sroa.2.0._arg_accA2.ascast.sroa_idx = getelementptr inbounds i8, ptr addrspace(4) %_arg_accA2.ascast, i64 8
  +  %agg.tmp5.sroa.0.sroa.2.0.copyload = load i64, ptr addrspace(4) %agg.tmp5.sroa.0.sroa.2.0._arg_accA2.ascast.sroa_idx, align 8
  +  %agg.tmp6.sroa.0.sroa.0.0.copyload = load i64, ptr addrspace(4) %_arg_accA3.ascast, align 8
  +  %agg.tmp6.sroa.0.sroa.2.0._arg_accA3.ascast.sroa_idx = getelementptr inbounds i8, ptr addrspace(4) %_arg_accA3.ascast, i64 8
  +  %agg.tmp6.sroa.0.sroa.2.0.copyload = load i64, ptr addrspace(4) %agg.tmp6.sroa.0.sroa.2.0._arg_accA3.ascast.sroa_idx, align 8
     %0 = load i64, ptr addrspace(4) getelementptr inbounds (i8, ptr addrspace(4) addrspacecast (ptr addrspace(1) @__spirv_BuiltInGlobalInvocationId to ptr addrspace(4)), i64 8), align 8, !noalias !123
     %1 = load i64, ptr addrspace(4) addrspacecast (ptr addrspace(1) @__spirv_BuiltInGlobalInvocationId to ptr addrspace(4)), align 32, !noalias !123
     %2 = load i64, ptr addrspace(4) getelementptr inbounds (i8, ptr addrspace(4) addrspacecast (ptr addrspace(1) @__spirv_BuiltInLocalInvocationId to ptr addrspace(4)), i64 8), align 8, !noalias !130
  @@ -43,10 +50,16 @@
     br label %for.cond.i.i, !llvm.loop !135

   _ZZZ19matrix_verify_logicIaLm16ELm16EEvN4sycl3_V15queueER10big_matrixIT_XT0_EXT1_EERNS1_8nd_rangeILi2EEEiENKUlRNS1_7handlerEE_clESB_ENKUlNS1_7nd_itemILi2EEEE_clESE_.exit: ; preds = %for.cond.i.i
  +  %mul.i6.i.i.i.i = mul i64 %agg.tmp6.sroa.0.sroa.0.0.copyload, %agg.tmp5.sroa.0.sroa.2.0.copyload
  +  %4 = getelementptr i8, ptr addrspace(1) %_arg_accA, i64 %mul.i6.i.i.i.i
  +  %add.ptr.i = getelementptr i8, ptr addrspace(1) %4, i64 %agg.tmp6.sroa.0.sroa.2.0.copyload
     %sub5.i = sub nsw i64 %1, %3
     %sub.i = sub nsw i64 %0, %2
  +  %add.i7.i.i.i.i.i.i = add i64 %mul.i6.i.i.i.i, %agg.tmp6.sroa.0.sroa.2.0.copyload
  +  %idx.neg.i.i.i = sub i64 0, %add.i7.i.i.i.i.i.i
  +  %add.ptr.i.i.i = getelementptr inbounds i8, ptr addrspace(1) %add.ptr.i, i64 %idx.neg.i.i.i
     %mul13.i = shl nsw i64 %sub.i, 9
  -  %add.ptr.i.i = getelementptr inbounds i8, ptr addrspace(1) %_arg_accA, i64 %mul13.i
  +  %add.ptr.i.i = getelementptr inbounds i8, ptr addrspace(1) %add.ptr.i.i.i, i64 %mul13.i
     %div16.i = shl nsw i64 %sub5.i, 2
     %mul15.i = and i64 %div16.i, -32
     %add.ptr.i30.i = getelementptr inbounds i8, ptr addrspace(1) %add.ptr.i.i, i64 %mul15.i

```

**Comment by jsji** [comment]
https://github.com/intel/llvm/pull/11521 will revert https://github.com/intel/llvm/commit/e13bed4c5f3544c076ce57e36d9a11eefa5a7815 temporarily for now. We need to find the solution and reland it ASAP. Thanks.

I assign to @YuriPlyakhin first for now, feel free to reassign to corresponding owners. Thanks.
FYI. @dkhaldi

**Comment by dkhaldi** [comment]
Adding @MrSidims

**Comment by bader** [comment]
I can't reproduce this issue on the tip of the `sycl` branch. See https://github.com/intel/llvm/pull/13243 pre-commit results. There were 10 attempts, but none of them failed on timeout. They fail due to another bug.

I'm going to re-enable disabled tests and track their status in CI.

**Comment by dkhaldi** [comment]
@bader, the issue is only reproducible on the CI machine as it is 750 DG2 hardware. I would assume the machine you were trying on is 770, that's why you cannot reproduce it.

**Comment by bader** [comment]
> @bader, the issue is only reproducible on the CI machine as it is 750 DG2 hardware. I would assume the machine you were trying on is 770, that's why you cannot reproduce it.

CI still uses "Intel(R) Arc(TM) A750 Graphics". You can confirm that by logs for PR I mentioned in my previous comment.

**Comment by YuriPlyakhin** [comment]
@bader ,
This test was also disabled in pre-check-in by this issue: https://github.com/intel/llvm/blob/27a3990dd7dce0614976c77e44efb05231a31211/sycl/test-e2e/Matrix/element_wise_all_ops.cpp#L11
So, it needs to be enabled as well

**Comment by bader** [comment]
> @bader , This test was also disabled in pre-check-in by this issue:
> 
> https://github.com/intel/llvm/blob/27a3990dd7dce0614976c77e44efb05231a31211/sycl/test-e2e/Matrix/element_wise_all_ops.cpp#L11
> 
> 
> So, it needs to be enabled as well

@YuriPlyakhin, this issue is about XMX8 tests only. Are you saying that test-e2e/Matrix/element_wise_all_ops.cpp was impacted by the same problem?

**Comment by YuriPlyakhin** [comment]
@bader , it is still about XMX8 test. `SYCL :: Matrix/XMX8/element_wise_all_ops.cpp` was disabled in this issue. Later it was replaced by code update in `SYCL :: Matrix/element_wise_all_ops.cpp`, which now supports also XMX8. So, before `SYCL :: Matrix/XMX8/element_wise_all_ops.cpp` was used to test XMX8, then it was disabled, then replaced by code change to use `SYCL :: Matrix/element_wise_all_ops.cpp` to test XMX8 part, but `UNSUPPORTED` tag was moved for dg2 to `SYCL :: Matrix/element_wise_all_ops.cpp` to keep XMX8 part disabled.

**Comment by bader** [comment]
> Later it was replaced by code update in `SYCL :: Matrix/element_wise_all_ops.cpp`, which now supports also XMX8. 

Does it mean `SYCL :: Matrix/XMX8/element_wise_all_ops.cpp` duplicates `SYCL :: Matrix/element_wise_all_ops.cpp` and can be removed?

> So, initially, `SYCL :: Matrix/XMX8/element_wise_all_ops.cpp` was used to test XMX8, then it was disabled, then replaced by code change to use `SYCL :: Matrix/element_wise_all_ops.cpp` to test XMX8 part, and `UNSUPPORTED` tag for dg2 was moved to `SYCL :: Matrix/element_wise_all_ops.cpp` to inherit keeping XMX8 part disabled.

Do you believe that it was the only issue with `SYCL :: Matrix/element_wise_all_ops.cpp`?

**Comment by bader** [comment]
@YuriPlyakhin, https://github.com/intel/llvm/pull/13323 enables `sycl/test-e2e/Matrix/element_wise_all_ops.cpp`.

**Comment by YuriPlyakhin** [comment]
> Do you believe that it was the only issue with `SYCL :: Matrix/element_wise_all_ops.cpp`?

That was the only issue related to XMX8.

**Comment by bader** [comment]
Okay. I merged #13323. Let's track the status of enabled tests in CI.
So far there were no Matrix test failures in post-commit since they were enabled.
